File: go-implementation/gameoflife/sequential_game.go
package main

import (
	"encoding/csv"
	"fmt"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
	"time"
)

func newMatrix(r, c int) [][]int {
	a := make([]int, c*r)
	m := make([][]int, r)
	lo, hi := 0, c
	for i := range m {
		m[i] = a[lo:hi:hi]
		lo, hi = hi, hi+c
	}
	return m
}

func loadMatrixFromFile(matrix_path string, n int) [][]int {
	matrix := newMatrix(n, n)

	content, err := ioutil.ReadFile(matrix_path)

	if err != nil {
		log.Fatal(err)
	}

	file := strings.NewReader(string(content))
	lines, _ := csv.NewReader(file).ReadAll()
	for i, line := range lines {
		for j, val := range line {
			val_int, err := strconv.Atoi(val)
			if err != nil {
				log.Fatal(err)
			}
			matrix[i][j] = val_int
		}
	}

	return matrix

}

func printMatrix(matrix [][]int, n int) {
	if true {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				fmt.Print(fmt.Sprint(matrix[i][j]) + " ")
			}
			fmt.Println()
		}
		fmt.Println()
	}
}

func copyMatrix(world [][]int, newWorld [][]int, n int) {
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			world[i][j] = newWorld[i][j]
		}
	}
}

func getNeighbour(row int, col int, world [][]int, n int) int {
	// If a neighbour of a cell is outside the limits of the world it is considered dead.
	if row >= n || row < 0 || col >= n || col < 0 {
		return 0
	}

	return world[row][col]
}

func getNeighboursCount(row int, col int, world [][]int, n int) int {
	count := 0

	count += getNeighbour(row-1, col-1, world, n)
	count += getNeighbour(row-1, col, world, n)
	count += getNeighbour(row-1, col+1, world, n)

	count += getNeighbour(row, col-1, world, n)
	count += getNeighbour(row, col+1, world, n)

	count += getNeighbour(row+1, col-1, world, n)
	count += getNeighbour(row+1, col, world, n)
	count += getNeighbour(row+1, col+1, world, n)

	return count
}

func rule(world [][]int, row int, col int, n int) int {
	count := getNeighboursCount(row, col, world, n)
	birth := world[row][col] == 0 && count == 3
	survive := world[row][col] == 1 && (count == 2 || count == 3)
	if birth || survive {
		return 1
	}
	return 0
}

func applyRules(world [][]int, newWorld [][]int, n int) {
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			newWorld[i][j] = rule(world, i, j, n)
		}
	}
}

func main() {
	n := 7
	m := 800000

	world := loadMatrixFromFile("resources/infinite_growth_1.txt", n)
	newWorld := newMatrix(n, n)

	fmt.Println("World")
	printMatrix(world, n)
	start := time.Now()

	for i := 0; i < m; i++ {
		applyRules(world, newWorld, n)
		copyMatrix(world, newWorld, n)
		// printMatrix(world, n)
	}

	fmt.Println("Last generation")
	printMatrix(world, n)

	elapsed := time.Since(start)
	fmt.Printf("Sequential game of life took: %d milliseconds", elapsed.Milliseconds())
}

File: python-implementation/gameoflife/sequential_game.py
import numpy as np
import time


def load_matrix_from_file(matrix_path):
    result_list = []
    f = open(matrix_path, 'r')
    for line in f:
        line = line.rstrip('\n') 
        string_value = line.split(',')
        int_value = list(map(np.int, string_value))
        result_list.append(int_value)
    f.close()

    print('Done loading ' + matrix_path)

    return result_list


def copyMatrix(world, new_world, n):
    for i in range(len(world[0])):
        for j in range(len(world)):
            world[i][j] = new_world[i][j]


def get_neighbour(row, col, world, n):
    if (row >= n or row < 0 or col >= n or col < 0):
        return 0

    return world[row][col]


def get_neighbours_count(row, col, world, n):
    count = 0

    count += get_neighbour(row-1, col-1, world, n)
    count += get_neighbour(row-1, col, world, n)
    count += get_neighbour(row-1, col+1, world, n)

    count += get_neighbour(row, col-1, world, n)
    count += get_neighbour(row, col+1, world, n)

    count += get_neighbour(row+1, col-1, world, n)
    count += get_neighbour(row+1, col, world, n)
    count += get_neighbour(row+1, col+1, world, n)

    return count


def rule(world, row, col, n):
    count = get_neighbours_count(row, col, world, n)
    birth = world[row][col] == 0 and count == 3
    survive = world[row][col] == 1 and (count == 2 or count == 3)
    if (birth or survive):
      return 1
    return 0


def apply_rules(world, new_world, n):
    for i in range(len(world[0])):
      for j in range(len(world)):
        new_world[i][j] = rule(world, i, j, n)
      
    
def print_matrix(matrix):
    if True:
        print(matrix)


def main():
    n = 28
    m = 20000

    world = load_matrix_from_file('resources/infinite_growth_1.txt')
    new_world = np.zeros((n, n)).astype(int)
    print("World")
    print_matrix(world)

    start = time.perf_counter()
    for i in range(m):
        apply_rules(world, new_world, n)
        copyMatrix(world, new_world, n)
    end = time.perf_counter()
    
    print("Last generation")
    print_matrix(world)

    print(f"Sequential game of life took: {round(end - start, 5)} seconds")

main()